### PR TITLE
Enhancement Issue #5737 "Router 2.5.4 multiple extensions in route"

### PR DESCRIPTION
**Pull request by Kwaadpepper**
https://github.com/cakephp/cakephp/issues/5737

Changed ext parse in Router where it was copying the 'ext' arg of
Router::connect.

So if you gave array('rss','atom') then the array was putted in controller->ext
instead of "choosing the correct ext in the array"

### CODE REVIEW

File: lib/Cake/Routing/Router.php [hunk 1]
```diff
@@ -619,7 +619,7 @@ public static function parse($url) {
 			$out['action'] = $out['prefix'] . '_' . $out['action'];
 		}
 
-		if (!empty($ext) && !isset($out['ext'])) {
+		if ((!empty($ext) && !isset($out['ext'])) || (gettype($out['ext']) == 'array')) {
```
Comment (dereuromark) [inline lib/Cake/Routing/Router.php]: isnt `(gettype($out['ext']) == 'array')` the same as just `is_array(...)`?

Comment (Kwaadpepper) [inline lib/Cake/Routing/Router.php]: yes, i think i just gonna sleep now, i can't do anything good x(

File: lib/Cake/Routing/Router.php [hunk 2]
```diff
@@ -619,8 +619,12 @@ public static function parse($url) {
 			$out['action'] = $out['prefix'] . '_' . $out['action'];
 		}
 
-		if (!empty($ext) && !isset($out['ext'])) {
-			$out['ext'] = $ext;
+		if (!empty($ext)) {
+			if (!isset($out['ext'])) {
+				$out['ext'] = $ext;
+			} elseif (is_array($out['ext'])) {
+				$out['ext'] = $ext;
```
Comment (markstory) [inline lib/Cake/Routing/Router.php]: These two conditions do the same thing, so the branches can probably be collapsed.

Comment (thaJeztah) [inline lib/Cake/Routing/Router.php]: Would it make sense to normalize and *always* use an array? e.g. `$out['ext'] = (array) $ext` (Some tests will have to be modified for that)

**Comment by Kwaadpepper** [comment]
while the pull request works for
```
Router::connect('/rss/:fluxname/*',array('controller' => 'misc','action' => 'feeds','ext' => array('rss','atom'));
```

i just saw that for 
```
url : http://myproject/rss/feeds.atom

Router::connect('/rss/:fluxname/*',array('controller' => 'misc','action' => 'feeds','ext' => 'rss'));
Router::connect('/rss/:fluxname/*',array('controller' => 'misc','action' => 'feeds','ext' => 'atom'));
```

i always get the first route to match, (tested with cake 2.5.1 and 2.5.6)

**Comment by thaJeztah** [comment]
> i always get the first route to match, (tested with cake 2.5.1 and 2.5.6)

Not sure what the *expected* outcome should be here (second overwriting the first, both routes being added?)

Perhaps `ext` should always use an array and do an array_unique when adding/setting extensions?

**Comment by Kwaadpepper** [comment]
> Not sure what the expected outcome should be here (second overwriting the first, both routes being added?)

I expect to:
url : http://myproject/rss/feeds.rss => match to route 1
url : http://myproject/rss/feeds.atom => match to route 2
url : http://myproject/rss/feeds.xml => lead to 404

As a work around i got this

```
diff --git a/Routing/Router.php b/Routing/Router.php
index 27fd10d..e2a2678 100755
--- a/Routing/Router.php
+++ b/Routing/Router.php
@@ -612,8 +612,8 @@ class Router {
 
                for ($i = 0, $len = count(self::$routes); $i < $len; $i++) {
                        $route =& self::$routes[$i];
-
-                       if (($r = $route->parse($url)) !== false) {
+                       $rext = isset($route->defaults['ext']) ? $route->defaults['ext'] : null;
+                       if (($r = $route->parse($url)) !== false && $rext == $ext) {
                                self::$_currentRoute[] =& $route;
                                $out = $r;
                                break;
```

> Perhaps ext should always use an array and do an array_unique when adding/setting extensions?
I'm not sure how to do this, i'm still looking for the changes it would takes

**Comment by markstory** [comment]
@Kwaadpepper If the example with two routes worked, would these changes be necessary? The exampl with two routes should work. It not working is an issue.

**Comment by Kwaadpepper** [comment]
Well the pull request works to use Router::connect with 'ext' => array('atom','ext').

But the router still fails to identify an url to match the route with the correct extension if you have 2 same routes but with different extensions.

It turns out that my pull request doesn't work with the patch in the previous post.

**Comment by Kwaadpepper** [comment]
Now both routes style are working with that (multiple and array), But the test fails whereas it works on my project, i may be missing something any ideas ?

```
index 0acde4f..bdef131 100644
--- a/lib/Cake/Routing/Router.php
+++ b/lib/Cake/Routing/Router.php
@@ -608,9 +608,14 @@ class Router {
 
 		extract(self::_parseExtension($url));
 
-		foreach (self::$routes as $route) {
-			if (($r = $route->parse($url)) !== false) {
-				self::$_currentRoute[] = $route;
+		for ($i = 0, $len = count(self::$routes); $i < $len; $i++) {
+			$route =& self::$routes[$i];
+			$rext = isset($route->defaults['ext']) ? $route->defaults['ext'] : null;
+			if (is_array($rext)) {
+				$rext = in_array($ext, $rext) ? $ext : null;
+			}
+			if (($r = $route->parse($url)) !== false && $rext == $ext) {
+				self::$_currentRoute[] =& $route;
 				$out = $r;
 				break;
 			}
@@ -619,8 +624,12 @@ class Router {
 			$out['action'] = $out['prefix'] . '_' . $out['action'];
 		}
 
-		if (!empty($ext) && (!isset($out['ext']) || gettype($out['ext']))) {
-			$out['ext'] = $ext;
+		if (!empty($ext)) {
+			if (!isset($out['ext'])) {
+				$out['ext'] = $ext;
+			} elseif (is_array($out['ext'])) {
+				$out['ext'] = $ext;
+			}
 		}
 
 		if (!empty($queryParameters) && !isset($out['?'])) {
diff --git a/lib/Cake/Test/Case/Routing/RouterTest.php b/lib/Cake/Test/Case/Routing/RouterTest.php
index 93394fa..082d997 100644
--- a/lib/Cake/Test/Case/Routing/RouterTest.php
+++ b/lib/Cake/Test/Case/Routing/RouterTest.php
@@ -1465,6 +1465,17 @@ class RouterTest extends CakeTestCase {
 
 		Router::reload();
 		Router::parseExtensions('rss', 'atom');
+		Router::connect('/controller/action', array('controller' => 'controller', 'action' => 'action', 'ext' => 'rss'));
+		Router::connect('/controller/action', array('controller' => 'controller', 'action' => 'action', 'ext' => 'atom'));
+		$result = Router::parse('/controller/action.rss');
+		$expected = array('controller' => 'controller', 'action' => 'action', 'plugin' => null, 'ext' => 'rss', 'named' => array(), 'pass' => array());
+		$this->assertEquals($expected, $result);
+		$result = Router::parse('/controller/action.atom');
+		$expected = array('controller' => 'controller', 'action' => 'action', 'plugin' => null, 'ext' => 'atom', 'named' => array(), 'pass' => array());
+		$this->assertEquals($expected, $result);
+
+		Router::reload();
+		Router::parseExtensions('rss', 'atom');
 		Router::connect('/controller/action', array('controller' => 'controller', 'action' => 'action', 'ext' => array('rss', 'atom')));
 		$result = Router::parse('/controller/action.rss');
 		$expected = array('controller' => 'controller', 'action' => 'action', 'plugin' => null, 'ext' => 'rss', 'named' => array(), 'pass' => array());

```

**Comment by Kwaadpepper** [comment]
My current solution for both problems, it works on my project but tests fails, any idea ?
```
diff --git a/lib/Cake/Routing/Router.php b/lib/Cake/Routing/Router.php
index 0acde4f..2f8bca5 100644
--- a/lib/Cake/Routing/Router.php
+++ b/lib/Cake/Routing/Router.php
@@ -609,18 +609,27 @@ class Router {
 		extract(self::_parseExtension($url));
 
 		foreach (self::$routes as $route) {
-			if (($r = $route->parse($url)) !== false) {
+			$rext = isset($route->defaults['ext']) ? $route->defaults['ext'] : null;
+			if(is_array($rext)) {
+				$rext = in_array($ext, $rext) ? $ext : null;
+			}
+			if (($r = $route->parse($url)) !== false && $rext == $ext) {
 				self::$_currentRoute[] = $route;
 				$out = $r;
 				break;
 			}
 		}
+
 		if (isset($out['prefix'])) {
 			$out['action'] = $out['prefix'] . '_' . $out['action'];
 		}
 
-		if (!empty($ext) && (!isset($out['ext']) || gettype($out['ext']))) {
-			$out['ext'] = $ext;
+		if (!empty($ext)) {
+			if (!isset($out['ext'])) {
+				$out['ext'] = $ext;
+			} elseif (is_array($out['ext'])) {
+				$out['ext'] = $ext;
+			}
 		}
 
 		if (!empty($queryParameters) && !isset($out['?'])) {
diff --git a/lib/Cake/Test/Case/Routing/RouterTest.php b/lib/Cake/Test/Case/Routing/RouterTest.php
index 93394fa..082d997 100644
--- a/lib/Cake/Test/Case/Routing/RouterTest.php
+++ b/lib/Cake/Test/Case/Routing/RouterTest.php
@@ -1465,6 +1465,17 @@ class RouterTest extends CakeTestCase {
 
 		Router::reload();
 		Router::parseExtensions('rss', 'atom');
+		Router::connect('/controller/action', array('controller' => 'controller', 'action' => 'action', 'ext' => 'rss'));
+		Router::connect('/controller/action', array('controller' => 'controller', 'action' => 'action', 'ext' => 'atom'));
+		$result = Router::parse('/controller/action.rss');
+		$expected = array('controller' => 'controller', 'action' => 'action', 'plugin' => null, 'ext' => 'rss', 'named' => array(), 'pass' => array());
+		$this->assertEquals($expected, $result);
+		$result = Router::parse('/controller/action.atom');
+		$expected = array('controller' => 'controller', 'action' => 'action', 'plugin' => null, 'ext' => 'atom', 'named' => array(), 'pass' => array());
+		$this->assertEquals($expected, $result);
+
+		Router::reload();
+		Router::parseExtensions('rss', 'atom');
 		Router::connect('/controller/action', array('controller' => 'controller', 'action' => 'action', 'ext' => array('rss', 'atom')));
 		$result = Router::parse('/controller/action.rss');
 		$expected = array('controller' => 'controller', 'action' => 'action', 'plugin' => null, 'ext' => 'rss', 'named' => array(), 'pass' => array());

```

**Comment by thaJeztah** [comment]
> But i guess it would be useless to support multiple routes with different extensions if you can give an array?

Not sure; having both options could be useful. Also if that works now, and no longer after this PR, it will be a BC breaking change, which cannot be in a .x release.

**Comment by Kwaadpepper** [comment]
Simple problem simple solution i hope.

**Comment by markstory** [comment]
I would prefer to avoid having more micro syntaxes. I can look into fixing that issue. As for restricting the extension, have you tried putting the extension in the route template?

**Comment by Kwaadpepper** [comment]
It doest work,
```
Router::connect('/rss.rss',array('controller' => 'misc','action' => 'feeds'));
```
throws 404 whereas
```
Router::connect('/rss.rss',array('controller' => 'misc','action' => 'feeds'));
```
this is because matching test doesn't use the extension, so if you give an extension in template it would be a part of url and not ext.

in CakeRoute::parse
```
$this->_compiledRoute => string '#^/rss\.rss[/]*$#' (length=17)

urldecode($url) => string '/rss' (length=4)

```

**Comment by markstory** [comment]
@Kwaadpepper This resolved as I expected it:

```php
Router::connect('/tasks.rss', ['controller' => 'tasks', 'action' => 'index', 'ext' => 'rss']);
```

This allows `/tasks.rss` to work as expected with the correct `ext` param, and `/tasks.atom` was a 404 (as it was not connected). Adding an additional route for `/tasks.atom` made that route resolve as well.

Does the above solve your original issue?

**Comment by Kwaadpepper** [comment]
I just tried with my cake2 project and it fails,
```
Router::connect('/task.rss', array('controller' => 'controller', 'action' => 'action', 'ext' => 'rss'));
		$result = Router::parse('task.rss');
		$expected = array('controller' => 'controller', 'action' => 'action', 'plugin' => null, 'ext' => 'rss', 'named' => array(), 'pass' => array());
		$this->assertEquals($expected, $result);
```

fails on last version of cake2.

If you were talking about using this tip on cake3 then i guess it is ok to close the PR.

**Comment by Kwaadpepper** [comment]
On cake2 your tips works with 
```
Router::parseExtensions('rss', 'xml');

Router::connect('/task.rss', array('controller' => 'controller', 'action' => 'action', 'ext' => 'rss'));
$result = Router::parse('/tasks.rss');
$expected = array('controller' => 'tasks', 'action' => 'index', 'plugin' => null, 'ext' => 'rss', 'named' => array(), 'pass' => array());
$this->assertEquals($expected, $result);
```
So can't choose controller or action, maybe you were talking about cakephp3

**Comment by markstory** [comment]
In your example you connect `/task.rss` and parse `/tasks.rss` which are not the same is that a typo?

I wasn't talking about Cake3. The quick check I did was against 2.6.1, and I didn't use `parseExtensions()` as that enables extensions on _all_ routes which doesn't seem to be something you wanted as you're trying to limit the extensions that apply to specific routes - something that is not possible with `parseExtensions()`.

**Comment by Kwaadpepper** [comment]
@markstory 

>In your example you connect /task.rss and parse /tasks.rss which are not the same is that a typo?

yes, but it doesn't matter since :test is given in action 

Ok for the parse extension thing, i always use it maybe i shouldn't.

Generally in my apps I serve ajax answers so i use parseExtensions(), and because i like the automatic behavior of serializing.

If i don't use parseExtensions(), I will loose the _serialize option for xml and json ?
I think that if i give 'ext' => 'json' the _serialize bahavior will work no ?
If that is the case than yes it solves my problem.

**Comment by markstory** [comment]
The `_serialize` features are not provided by `Router`, they are provided by `RequestHandlerComponent`. So you won't lose that functionality if you declare routes with `ext` and don't use `parseExtensions()`.

**Comment by Kwaadpepper** [comment]
Ok thanks, problem solved

**Pull request closed without merging**